apps/namada-interface/src/slices/accounts.ts

```diff
@@ -135,3 +143,55 @@ const { actions, reducer } = accountsSlice;
 
 export const { addAccounts } = actions;
 export default reducer;
+
+////////////////////////////////////////////////////////////////////////////////
+// JOTAI
+////////////////////////////////////////////////////////////////////////////////
+
+const accountsAtom = (() => {
+  const base = atom(new Promise<readonly AccountDetails[]>(() => {}));
+
+  return atom(
+    (get) => get(base),
+    async (_get, set) => {
+      const namada = getIntegration("namada");
+      const accounts = namada.accounts().then(async (accounts) => {
+        if (accounts === undefined) {
+          throw new Error("accounts was undefined!");
+        }
+        return accounts;
+      });
+      set(base, accounts);
+    }
+  );
+})();
+
+const balancesAtom = (() => {
+  const base = atom<Record<Address, Balance>>({});
+
+  return atom(
+    (get) => get(base),
+    async (get, set) => {
+      const accounts = await get(accountsAtom);
+      const namada = getIntegration("namada");
+
+      // TODO: should we be throwing away old balances here?
+      const balances = get(base);
+
+      accounts.forEach(async (account) => {
+        const result = await namada.queryBalances(account.address);
+        const balance = result.reduce(
+          (acc, curr) => ({ ...acc, [curr.token]: new BigNumber(curr.amount) }),
+          {} as Balance
+        );
+        balances[account.address] = balance;
+        set(
+          base,
+          { ...balances } // object clone ensures jotai realises object has changed
```

Comment:
Not sure how helpful that would be in this case but there is an [immer extension](https://jotai.org/docs/extensions/immer) for jotai :)